website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,826 @@
+
+    // TODO: better extensions system?
+    self.extensions = {
+        'osf-upload': editorExtensions.Uploader
+    };
+};
+/**
+ * Load draft data into the editor
+ *
+ * @param {Draft} draft
+ **/
+RegistrationEditor.prototype.init = function(draft) {
+    var self = this;
+
+    self.draft(draft);
+    var metaSchema = draft.metaSchema;
+
+    var schemaData = {};
+    if (draft) {
+        schemaData = draft.schemaData || {};
+    }
+
+    // Set currentPage to the first page
+    self.currentPage(self.draft().pages()[0]);
+
+    self.serialized.subscribe(self.autoSave.bind(self));
+};
+/**
+ * @returns {Question[]} flat list of the current schema's questions
+ **/
+RegistrationEditor.prototype.flatQuestions = function() {
+    var self = this;
+
+    return self.draft().metaSchema.flatQuestions();
+};
+/**
+ * Creates a template context for editor type subtemplates. Looks for the data type
+ * in the extension map, and if found instantiates that type's ViewModel. Otherwise
+ * return the bare data Object
+ *
+ * @param {Object} data: data in current editor template scope
+ * @returns {Object|ViewModel}
+ **/
+RegistrationEditor.prototype.context = function(data) {
+    $.extend(data, {
+        save: this.save.bind(this),
+        readonly: this.readonly
+    });
+
+    if (this.extensions[data.type]) {
+        return new this.extensions[data.type](data);
+    }
+    return data;
+};
+/**
+ * Check that the Draft is valid before registering
+ */
+RegistrationEditor.prototype.check = function() {
+    var self = this;
+
+    ko.utils.arrayMap(self.draft().pages(), function(page) {
+        ko.utils.arrayMap(page.questions(), function(question) {
+            if (question.required && !question.value.isValid()) {
+                // Validation for a question failed
+                bootbox.dialog({
+                    title: 'Registration Not Complete',
+                    message: 'There are errors in your registration. Please double check it and submit again.',
+                    buttons: {
+                        success: {
+                            label: 'Ok',
+                            className: 'btn-success',
+                            callback: function() {
+                                self.showValidation(true);
+                            }
+                        }
+                    }
+                });
+                return;
+            }
+            // Validation passed for all applicable questions
+            window.location = self.draft().urls.register_page;
+        });
+    });
+};
+/**
+ * Select a page, selecting the first question on that page
+ **/
+RegistrationEditor.prototype.selectPage = function(page) {
+    var self = this;
+
+    // var firstQuestion = page.questions[Object.keys(page.questions)[0]];
+    self.currentPage(page);
+};
+
+RegistrationEditor.prototype.nextPage = function () {
+    var self = this;
+    if (self.onLastPage() || self.pages().length < 2) {
```

Comment:
Given that a 0-page document seems unlikely, would it be fair to say this is equivalent to `self.pages().length === 1` ?

Be as explicit as possible when communicating intent.